### ISSUE TITLE
Automatic recover of last processing profile seems to fail in some cases

### ISSUE THREAD

**Issue by simohno** [Beep6581/RawTherapee#6098]
I like the fact that, when opening a file in Editor mode, RT seems to seek aside the file (or in the cache folder, depending on the setting about where to save the pp3), for a pp3 already present, named in the same way, and it loads it automatically to recover the last processing settings.

I noticed though that in some cases, when a file (e.g. in my case a developed RAW I am satisfied with) is moved to another folder (e.g. with File Operations > Move to...), and then it is moved back to its original folder (because I changed my mind and want to develop it in another way), RT loads the _default_ processing profile settings and saves a new pp3 with them. This, at least, is true for me which uses only the central cache folder to save and load the pp3s.
The solution for me is to click on _Load profile from file_ (above the tools palette) and select the pp3 saved BEFORE the last one (which as said includes default settings), this way settings are restored correctly (_except_, I don't know why, the color label).

This small bug seems to happen when the path of the file changes dramatically/the file is externally moved to the system recycle bin and then recovered. I tried moving temporarily a file to up one level in its path, and RT recovers from the cache the pp3 correctly.

Thanks for the investigation
Simone

RT version info:
_Version: 5.8-2741-ga5280fddf
Branch: dev
Commit: a5280fddf
Commit date: 2021-01-26
Compiler: x86_64-w64-mingw32-gcc\ 9.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.24.2
Lensfun: V0.3.2.0
Build type: Release
Build flags: -mwin32\ -m64\ -mthreads\ -msse2\ -std=c++11\ -mtune=generic\ -Werror=unused-label\ -Werror=delete-incomplete\ -fno-math-errno\ -Wall\ -Wuninitialized\ -Wcast-qual\ -Wno-deprecated-declarations\ -Wno-unused-result\ -Wunused-macros\ -fopenmp\ -Werror=unknown-pragmas\ -Wno-aggressive-loop-optimizations\ -DNDEBUG\ -O3\ -ftree-vectorize
Link flags: -m64\ -mthreads\ -static-libgcc\ -mtune=generic\ -s\ -O3
OpenMP support: ON
MMAP support: ON
Build OS: Linux 4.4.0-101-generic x86_64
Build date: Wed, 27 Jan 2021 01:24:35 +0000 UTC
Build epoch: 1611710675
Build UUID: 2ab8be1a-73f8-4a9d-a727-07f7f7f5fafd_

**Comment by ChasingShadows** [comment]
I'm using these settings

![image](https://user-images.githubusercontent.com/55600750/107762844-ebaef600-6d2d-11eb-97fa-101803642692.png)

adn always move the pp3 together with the raw. In addition. I have a pp3 for each output file.

I guess in your case to trust the cache ..

**Comment by simohno** [comment]
Possibly the problem is absent if you set up the software like that, and move always the RAW with its pp3 aside, because RT knows it has to search for the pp3 always next to the file (?)
The thing is that I personally don't like having those white files between the photo files in Explorer, the cache is far more neat system for me. And one could always access it via Run (in Windows) > _AppData\Local\RawTherapee5\cache_

**Comment by simohno** [comment]
From the 1st answer to #6110 I just got an idea of what the problem could be when restoring files to a folder: RT thinks that the restored file is a _new_ and _different_ file with identical name to another, already cached, and so tries not to misunderstand the former with the latter, whereas the file is really the same. Hence, as I verified, RT creates a new default pp3 with another casual suffix after the filename, and the processing settings are lost unless manually loading the old pp3.

Just an idea, and possibly not simple to resolve. There should be another criterion apart the name, to decide whether the file is really the same as that the pp3 is referring to: I'd suggest the exact file size.

The mystery is that in a few cases, as I said, the processing settings are recovered when the file is moved back to its original folder.

**Comment by ChasingShadows** [comment]
As far as I know, the criteria is a checksum. if the pp3 is not next to the file but in the cache.

If you move a file away and RT needs space in the cache, I'm pretty sure no longer existing files (from the cache point of view) lead to dump the related cache entry.  Hence, when you move the file to RT again, it doesn't find any saved pp3 file, not in the cache and not in the folder. So from the systems point of view this file has never been processed before.

**Comment by simohno** [comment]
The thing is that when I delete/move the file, no pp3 is deleted at the same time, I verified it keeping open the cache folder into AppData. When the file is moved back, a new pp3 (default settings) is generated with another suffix. The old, as I said, is still manually re-loadable.

Possibly it deletes an "entry" in another "internal" database? De-associating permanently that pp3
Or otherwise, there is something wrong with the checksum comparison, because I guess the file itself is not changed when deleted and restored only (if changed in any way, its checksum changes right?)

**Comment by Lawrence37** [comment]
@simohno Does RT fail to find the cached profile if you only move files using RT? The cache association should be kept when file operations are done in RT, otherwise it's a bug. I looked at the code and I believe the association is based on the file name and size (and creation time on Windows). If any of these change when you move the image using another program, RT won't be able to find the profile.

**Comment by simohno** [comment]
Hello Lawrence and thanks for replying

So, you mean that both the operations (moving and re-moving back) have to be done in RT (with _File operations_) for the cache association to be maintained? I will soon try with both behaviours (moving the file from Windows and from RT in both cases) and I'll let you know here.

Simone

**Comment by Lawrence37** [comment]
Moving images using RT guarantees the cache association is kept, but using another program also works as long as you ensure the file name, size, and creation time are kept.

**Comment by Beep6581** [comment]
https://github.com/Beep6581/RawTherapee/blob/dev/rtgui/cachemanager.cc#L310

**Comment by simohno** [comment]
Ok, I switched to RT version _dev-2781_ and tried moving a file from its original folder to desktop, using Windows Explorer, and with RT open, then moving it back. pp3 is correctly restored. The same if sending to the recycle bin.

But then I tried using RT (right menu > File operations > Move to...) to move to desktop and then immediately used Explorer to move it back, and the pp3 is not restored. So, strangely the bug seems to be present when one uses RT's tool "File operations" to move files to a folder, which I frequently use because I have one common folder for "Developed raws". Maybe that operation does something more than moving.
Can anyone reproduce this?

File name, size and creation time do not change in this operations, only the "Last access".

**Comment by Lawrence37** [comment]
Looks like the file name includes the full path. You can confirm this by applying a profile, moving the image using another program, applying another profile, then move the image back and forth between the two locations. RawTherapee will apply one of the two profiles depending on where the image is.

**Comment by simohno** [comment]
O.K. thus my suppositions are: 
- if I move a file with another program, I _don't reopen it_ in RT and move it back with another program, the full path in RT's database (or PP3 itself?) has not changed, so the old PP3 is restored
- if I conversely move the file with RT, it updates the file full path, so that if I move back the file with Explorer, the full path stays wrong and RT rejects the associated PP3 creating one new.
- if I always move from within RT the full paths should be updated in the PP3 every time

Summing up, one has to move the photos either always with RT (updating the full path) or always using Explorer, what I didn't use to do (I moved with RT and re-moved with explorer because it was quicker).

Does the above seem true to you? I will make a try in my next photo editing session. Thanks for the clarifications.
If this is confirmed, should we consider this behaviour a bug? I would expect that only the sole _file_ name is a criterion, but maybe checking the full path is intended.

**Comment by Lawrence37** [comment]
Correct.
Not sure if it's a bug. I can imagine that it's behaving as designed since you can have two images with the same name in different directories. If RT didn't use the full path, it would apply the same profile to both files, assuming they have the same size (could be common if the images are uncompressed). Regardless, I suppose it is possible to improve the cache association algorithm.
